neural_compressor/common/base_tune.py

```diff
@@ -0,0 +1,156 @@
+# limitations under the License.
+
+from abc import abstractmethod
+from typing import Any, Callable, Dict, List, Optional, Union
+
+from neural_compressor.common.base_config import BaseConfig, ComposableConfig
+from neural_compressor.common.logger import Logger
+
+logger = Logger().get_logger()
+
+
+class AlgorithmWrapper:
+    """Abstract base class for wrap algorithm implementation.
+
+    AlgorithmWrapper provides a uniform interface for encapsulating algorithm implementation
+    across different frameworks.
+    This class is intended to be used by a `tuner` to obtain quantized models.
+    """
+
+    def __init__(self, model) -> None:
+        self.model = model
+
+    @abstractmethod
+    def apply(self):
+        """The entry to apply algorithms on a given model."""
+        raise NotImplementedError
+
+
+class TuningObjective:
+    EVAL_FN = "eval_fn"
+    WEIGHT = "weight"
+    FN_NAME = "name"
+    EVAL_FN_TEMPLATE: Dict[str, Any] = {EVAL_FN: None, WEIGHT: 1.0, FN_NAME: None}
+
+    def __init__(self) -> None:
+        self.eval_fn_registry: List[Callable] = []
+
+    def evaluate(self, model) -> float:
+        """Evaluate the model using registered evaluation functions.
+
+        Args:
+            model: The fp32 model or quantized model.
+
+        Returns:
+            The overall result of all registered evaluation functions.
+        """
+        result = 0
+        for eval_pair in self.eval_fn_registry:
+            eval_fn = eval_pair[self.EVAL_FN]
+            eval_result = eval_fn(model)
+            result = self._update_the_objective_score(eval_pair, eval_result, result)
+        return result
+
+    def _update_the_objective_score(self, eval_pair, eval_result, overall_result):
+        # TODO update the result according to the weight and algo_name
+        return overall_result + eval_result * eval_pair[self.WEIGHT]
+
+    def get_number_of_tuning_objectives(self):
+        return len(self.eval_fn_registry)
+
+    def _update_eval_fn_registry(self, user_eval_fns: List[Dict]):
+        self.eval_fn_registry = [
+            {
+                self.EVAL_FN: user_eval_fn_pair[self.EVAL_FN],
+                self.WEIGHT: user_eval_fn_pair.get(self.WEIGHT, 1.0),
+                self.FN_NAME: user_eval_fn_pair.get(self.FN_NAME, user_eval_fn_pair[self.EVAL_FN].__name__),
+            }
+            for user_eval_fn_pair in user_eval_fns
+        ]
+
+    def update_eval_fn_registry(self, eval_fns: Optional[Union[Dict, List[Dict]]] = None) -> None:
+        if eval_fns is None:
+            return
+        elif isinstance(eval_fns, Dict):
+            eval_fns = [eval_fns]
+        elif isinstance(eval_fns, List):
+            assert all([isinstance(eval_fn_pair, Dict) for eval_fn_pair in eval_fns])
+        else:
+            raise NotImplementedError(f"The eval_fns should be a dict or a list of dict, but got {type(eval_fns)}.")
+        self._update_eval_fn_registry(eval_fns)
+
+
+tuning_objective = TuningObjective()
```

Comment:
define a default objective for use?

Reply:
The `TuningObjectives` comprises user-provided eval_fn, and there isn't a default value for it. The `TuningConfig` comes with a default value and will be added in the next PR.